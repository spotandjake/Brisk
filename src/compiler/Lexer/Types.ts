export interface Token {
  type: string;
  value: string | number | boolean | bigint;
  text: string;
  offset: number;
  line: number;
  col: number;
}
export interface Rule {
  type: string;
  id: string;
  value?: (text: string) => string | number | boolean | bigint;
  match: RegExp;
  lineBreaks?: boolean;
}