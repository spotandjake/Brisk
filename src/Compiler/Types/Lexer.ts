// Imports
import { Position } from '../Types/Types';
// Types
const enum LexemeType {
  keyword,
  separator,
  operator,
  literal,
  flag,
  comment,
  identifier
}
interface Lexeme {
  type: LexemeType;
  id: string;
  value?: (text: string) => string | number;
  match: RegExp;
  lineBreaks?: boolean;
}
interface Token {
  type: string;
  value: string | number;
  text: string;
  position: Position;
}
// Exports
export {
  LexemeType,
  Lexeme,
  Token
}