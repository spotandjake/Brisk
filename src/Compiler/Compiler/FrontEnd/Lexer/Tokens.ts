import { Lexeme, LexemeType } from '../../Types';
const tokens: Lexeme[] = [
  // keyword
  {
    type: LexemeType.keyword,
    id: 'Tkn_import',
    match: /import/
  },
  {
    type: LexemeType.keyword,
    id: 'Tkn_wasm',
    match: /wasm/
  },
  {
    type: LexemeType.keyword,
    id: 'Tkn_from',
    match: /from/
  },
  {
    type: LexemeType.keyword,
    id: 'Tkn_export',
    match: /export/
  },
  {
    type: LexemeType.keyword,
    id: 'Tkn_let',
    match: /let/
  },
  {
    type: LexemeType.keyword,
    id: 'Tkn_if',
    match: /if/
  },
  {
    type: LexemeType.keyword,
    id: 'Tkn_else',
    match: /else/
  },
  // separator
  {
    type: LexemeType.separator,
    id: 'Tkn_l_paren',
    match: /\(/
  },
  {
    type: LexemeType.separator,
    id: 'Tkn_r_paren',
    match: /\)/
  },
  {
    type: LexemeType.separator,
    id: 'Tkn_l_bracket',
    match: /\{/
  },
  {
    type: LexemeType.separator,
    id: 'Tkn_r_bracket',
    match: /\}/
  },
  {
    type: LexemeType.separator,
    id: 'Tkn_comma',
    match: /,/
  },
  {
    type: LexemeType.separator,
    id: 'Tkn_colon',
    match: /:/
  },
  {
    type: LexemeType.separator,
    id: 'Tkn_semicolon',
    match: /;/
  },
  {
    type: LexemeType.separator,
    id: 'Tkn_ws',
    match: /[ \t\s]+/,
    lineBreaks: true
  },
  // operator
  {
    type: LexemeType.operator,
    id: 'Tkn_arrow',
    match: /->/
  },
  {
    type: LexemeType.operator,
    id: 'Tkn_thick_arrow',
    match: /=>/
  },
  {
    type: LexemeType.operator,
    id: 'Tkn_equal',
    match: /=/
  },
  // literal
  {
    type: LexemeType.literal,
    id: 'Tkn_str',
    match: /'.*'/,
    value: (text: string): string => {
      return text.slice(1, text.length - 1)
        .replace(/(?<!\\)\\b/g, '\b')
        .replace(/(?<!\\)\\f/g, '\f')
        .replace(/(?<!\\)\\n/g, '\n')
        .replace(/(?<!\\)\\r/g, '\r')
        .replace(/(?<!\\)\\t/g, '\t')
        .replace(/(?<!\\)\\\\/g, '\\')
        // eslint-disable-next-line indent
        ;
    }
  },
  {
    type: LexemeType.literal,
    id: 'Tkn_number',
    match: /[-|+]?[0-9]*(?:\.?[0-9]+)/,
    value: (text: string): (number | bigint) => {
      if (Number(text) < 2147483647 && Number(text) > -2147483647 || !Number.isInteger(Number(text))) return Number(text);
      else return BigInt(text);
    }
  },
  {
    type: LexemeType.literal,
    id: 'Tkn_bool',
    match: /(?:true|false)/,
    value: (text: string): boolean => text == 'true'
  },
  // flag
  {
    type: LexemeType.flag,
    id: 'Tkn_flag',
    match: /@.*/,
    value: (text: string): string => text.slice(1)
  },
  // comment
  {
    type: LexemeType.comment,
    id: 'Tkn_comment',
    match: /\/\/.*/,
    value: (text: string): string => text.slice(2).trim()
  },
  // identifier
  {
    type: LexemeType.identifier,
    id: 'Tkn_identifier',
    match: /[a-zA-Z$_][1-9a-zA-Z$_]*/
  },
];

export default tokens;