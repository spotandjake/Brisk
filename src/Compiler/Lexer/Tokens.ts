import { Lexer, createToken } from 'chevrotain';
import { LexerTokenType } from '../Types/LexerNodes';
// =================================================================
// Categories
export const literalTokens = createToken({
  name: LexerTokenType.Literals,
  pattern: Lexer.NA
});
export const comparisonOperators = createToken({
  name: LexerTokenType.ComparisonOperators,
  pattern: Lexer.NA
});
export const assignmentOperators = createToken({
  name: LexerTokenType.AssignmentOperators,
  pattern: Lexer.NA
});
export const arithmeticOperators = createToken({
  name: LexerTokenType.ArithmeticOperators,
  pattern: Lexer.NA
});
export const logicalOperators = createToken({
  name: LexerTokenType.LogicalOperators,
  pattern: Lexer.NA
});
// Comments
export const TknComment = createToken({
  label: 'Comment',
  name: LexerTokenType.TknComment,
  pattern: /\/\/.*/
}); // Comment
// Keywords
export const TknImport = createToken({
  label: 'Import',
  name: LexerTokenType.ImportToken,
  pattern: /import/
});
export const TknWasm = createToken({
  label: 'Wasm Import Identifier',
  name: LexerTokenType.WasmToken,
  pattern: /wasm/
});
export const TknFrom = createToken({
  label: 'From',
  name: LexerTokenType.FromToken,
  pattern: /from/
});
export const TknExport = createToken({
  label: 'Export',
  name: LexerTokenType.ExportToken,
  pattern: /export/
});
export const TknConst = createToken({
  label: 'Const',
  name: LexerTokenType.ConstToken,
  pattern: /const/
});
export const TknLet = createToken({
  label: 'Let',
  name: LexerTokenType.LetToken,
  pattern: /let/
});
export const TknIf = createToken({
  label: 'If',
  name: LexerTokenType.IfToken,
  pattern: /if/
});
export const TknElse = createToken({
  label: 'Else',
  name: LexerTokenType.ElseToken,
  pattern: /else/
});
// Separators
export const TknLParen = createToken({
  label: 'Left Parenthesis',
  name: LexerTokenType.LeftParenthesis,
  pattern: /\(/
});
export const TknRParen = createToken({
  label: 'Right Parenthesis',
  name: LexerTokenType.RightParenthesis,
  pattern: /\)/
});
export const TknLBrace = createToken({
  label: 'Left Brace',
  name: LexerTokenType.LeftBrace,
  pattern: /\{/
});
export const TknRBrace = createToken({
  label: 'Right Brace',
  name: LexerTokenType.RightBrace,
  pattern: /\}/
});
export const TknLBracket = createToken({
  label: 'Left Bracket',
  name: LexerTokenType.LeftBracket,
  pattern: /\[/
});
export const TknRBracket = createToken({
  label: 'Right Bracket',
  name: LexerTokenType.RightBracket,
  pattern: /\]/
});
export const TknComma = createToken({
  label: 'Comma',
  name: LexerTokenType.Comma,
  pattern: /,/
});
export const TknPeriod = createToken({
  label: 'Period',
  name: LexerTokenType.Period,
  pattern: /\./
});
export const TknColon = createToken({
  label: 'Colon',
  name: LexerTokenType.Colon,
  pattern: /:/
});
export const TknSemiColon = createToken({
  label: 'Semicolon',
  name: LexerTokenType.Semicolon,
  pattern: /;/
});
export const TknWhitespace = createToken({
  label: 'Whitespace',
  name: LexerTokenType.WhiteSpace,
  pattern: /[ \t\s\r\n]+/,
  line_breaks: true
}); // ws
// Operators
export const TknComparisonEqual = createToken({
  label: 'Comparison Equal',
  name: LexerTokenType.TknComparisonEqual,
  categories: comparisonOperators,
  pattern: /==/
});
export const TknComparisonNotEqual = createToken({
  label: 'Comparison Not Equal',
  name: LexerTokenType.TknComparisonNotEqual,
  categories: comparisonOperators,
  pattern: /!=/
});
export const TknComparisonGreaterThan = createToken({
  label: 'Comparison Greater Than',
  name: LexerTokenType.TknComparisonGreaterThan,
  categories: comparisonOperators,
  pattern: />/
});
export const TknComparisonLessThan = createToken({
  label: 'Comparison Less Than',
  name: LexerTokenType.TknComparisonLessThan,
  categories: comparisonOperators,
  pattern: /</
});
export const TknComparisonGreaterThanEqual = createToken({
  label: 'Comparison Greater Than Or Equal To',
  name: LexerTokenType.TknComparisonGreaterThanOrEqual,
  categories: comparisonOperators,
  pattern: />=/
});
export const TknComparisonLessThanEqual = createToken({
  label: 'Comparison Less Than Or Equal To',
  name: LexerTokenType.TknComparisonLessThanOrEqual,
  categories: comparisonOperators,
  pattern: /<=/
});
export const TknThickArrow = createToken({
  label: 'Arrow',
  name: LexerTokenType.TknArrow,
  pattern: /=>/
});
export const TknThinArrow = createToken({
  label: 'Signature Arrow',
  name: LexerTokenType.TknThinArrow,
  pattern: /->/
});
export const TknNot = createToken({
  label: 'Not',
  name: LexerTokenType.TknNot,
  categories: logicalOperators,
  pattern: /!/
});
export const TknEqual = createToken({
  label: 'Equal',
  name: LexerTokenType.TknEqual,
  categories: assignmentOperators,
  pattern: /=/
});
export const TknAdd = createToken({
  label: 'Add',
  name: LexerTokenType.TknAdd,
  categories: arithmeticOperators,
  pattern: /\+/
});
export const TknSub = createToken({
  label: 'Subtract',
  name: LexerTokenType.TknSubtract,
  categories: arithmeticOperators,
  pattern: /-/
});
export const TknDiv = createToken({
  label: 'Division',
  name: LexerTokenType.TknDivision,
  categories: arithmeticOperators,
  pattern: /\//,
  longer_alt: TknComment
});
export const TknMul = createToken({
  label: 'Multiplication',
  name: LexerTokenType.TknMultiply,
  categories: arithmeticOperators,
  pattern: /\*/
});
export const TknPow = createToken({
  label: 'Power Of',
  name: LexerTokenType.TknPower,
  categories: arithmeticOperators,
  pattern: /\^/
});
// Literals
export const TknString = createToken({
  label: 'String Literal',
  name: LexerTokenType.TknStringLiteral,
  categories: literalTokens,
  pattern: /'.*'/
}); // String
export const TknNumber = createToken({
  label: 'Number Literal',
  name: LexerTokenType.TknNumberLiteral,
  categories: literalTokens,
  pattern: /[-|+]?[0-9]*(?:\.?[0-9]+)/
}); // Number
export const TknConstant = createToken({
  label: 'Constant Literal',
  name: LexerTokenType.TknConstantLiteral,
  categories: literalTokens,
  pattern: /(?:true|false|void)/
}); // Constant
export const TknWasmCall = createToken({
  label: 'Wasm Call',
  name: LexerTokenType.TknWasmCall,
  categories: literalTokens,
  pattern: /@wasm((\.[a-z\d]+)+)/
}); // Wasm Call
// Flags
export const TknFlag = createToken({
  label: 'Flag',
  name: LexerTokenType.TknFlag,
  pattern: /@.*/
}); // Flag
// Identifiers
export const TknIdentifier = createToken({
  label: 'Identifier',
  name: LexerTokenType.TknIdentifier,
  pattern: /[a-zA-Z$_][1-9a-zA-Z$_]*/
}); // Identifier
// =================================================================
export const Tokens = [
  // Categories
  literalTokens,
  comparisonOperators,
  assignmentOperators,
  arithmeticOperators,
  logicalOperators,
  // Keywords
  TknImport,
  TknWasm,
  TknFrom,
  TknExport,
  TknConst,
  TknLet,
  TknIf,
  TknElse,
  // Separators
  TknLParen,
  TknRParen,
  TknLBrace,
  TknRBrace,
  TknLBracket,
  TknRBracket,
  TknComma,
  TknPeriod,
  TknColon,
  TknSemiColon,
  TknWhitespace,
  TknComparisonEqual,
  TknComparisonNotEqual,
  TknComparisonGreaterThanEqual,
  TknComparisonLessThanEqual,
  TknComparisonGreaterThan,
  TknComparisonLessThan,
  TknThickArrow,
  TknThinArrow,
  TknNot,
  TknEqual,
  TknAdd,
  TknSub,
  TknDiv,
  TknMul,
  TknPow,
  // Literals
  literalTokens,
  TknString,
  TknNumber,
  TknConstant,
  TknWasmCall,
  // Flags
  TknFlag,
  // Comments
  TknComment,
  // Identifiers
  TknIdentifier,
];