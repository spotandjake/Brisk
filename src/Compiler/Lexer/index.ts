// Imports
import { Lexer } from 'chevrotain';
import { Tokens } from './Tokens';
import { BriskCustomError, BriskSyntaxError } from '../Errors/Compiler';
import { BriskErrorType } from '../Errors/Errors';
//@ts-ignore
import { __DEBUG__ } from '@brisk/config';
// =================================================================
// Lex code
const lex = (code: string, basePath: string, file: string) => {
  const tokenized = new Lexer(Tokens, {
    ensureOptimizations: true,
    skipValidations: !__DEBUG__,
  }).tokenize(code);
  if (tokenized.groups.hasOwnProperty('Reserved') && tokenized.groups.Reserved.length != 0)
    BriskSyntaxError(
      code,
      BriskErrorType.CannotUseReservedKeyword,
      [tokenized.groups.Reserved[0].image],
      {
        offset: tokenized.groups.Reserved[0].startOffset,
        length: tokenized.groups.Reserved[0].image.length,
        line: tokenized.groups.Reserved[0].startLine || 0,
        col: tokenized.groups.Reserved[0].startColumn || 0,
        basePath: basePath,
        file: file,
      }
    );
  if (tokenized.errors.length > 0) {
    const { offset, line, column, length, message } = tokenized.errors[0];
    const position = {
      offset: offset,
      length: length,
      line: line || 0,
      col: column || 0,
      basePath: basePath,
      file: file,
    };
    BriskCustomError(code, 'Syntax', message, position);
  }
  return tokenized;
};
// =================================================================
export default lex;
