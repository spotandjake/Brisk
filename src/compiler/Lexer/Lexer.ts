import { Rule, Token, Position } from './Types';
import Tokens from './Tokens';

class Lexer {
  private rules: Rule[] = Tokens;
  private position: Position = {
    offset: 0,
    line: 1,
    col: 0
  };
  lex(filename: string, rawCode: string): Token[] {
    const { rules, position } = this;
    const output: Token[] = [];
    let code: string = rawCode;
    // Lex Code
    while (code) {
      // Determine if the line is blank
      const line = code.split(/(?<=\r?\n)/)[0];
      if (line.trim().length == 0) {
        position.line++;
        position.offset += line.length;
        position.col = 0;
        code = code.slice(line.length);
        continue;
      }
      let foundMatch = false;
      // Match Rule
      for (const rule of rules) {
        const { type, id, value, match } = rule;
        const matchToken = code.match(match);
        if (matchToken) {
          foundMatch = true;
          // Update position
          position.offset += matchToken[0].length;
          position.col += matchToken[0].length;
          // Write Token to output
          const token: Token = {
            type: type,
            id: id,
            text: matchToken[0],
            value: value ? value(matchToken[0]) : matchToken[0].trim(),
            offset: position.offset-1,
            line: position.line,
            col: position.col
          };
          output.push(token);
          // Make input smaller
          code = code.slice(matchToken[0].length);
          break;
        }
      }
      if (!foundMatch) {
        // Throw A Syntax Error
        console.log(`SyntaxError: unexpected ${code[0]} at ${filename}:${position.line}:${position.col}`);
        process.exit(1);
      }
    }
    return output;
  }
}
export default Lexer;