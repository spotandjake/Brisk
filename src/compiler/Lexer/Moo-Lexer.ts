import Tokens from './Tokens';
// Import Types
import { Token, Rule } from './Types';
class Lexer {
  private regex: RegExp;
  private rules: Rule[];
  private dataset = '';
  private offset = 0;
  private line = 1;
  private col = 1;
  constructor(rules: Rule[]) {
    // Process rules
    rules.forEach(({ match }) => {
      if (match.ignoreCase) throw new Error('RegExp /i flag used');
      if (match.global) throw new Error('RegExp /g flag used');
      if (match.sticky) throw new Error('RegExp /y flag used');
      if (match.multiline) throw new Error('RegExp /m flag used');
    });
    // store stuff
    this.regex = new RegExp(`(?:${rules.map(({ match }) => `((?:${match.source}))`).join('|')})`, 'my');
    this.rules = rules;
  }
  next(): (Token | undefined) {
    const { regex, rules, dataset, offset, line, col } = this;
    // Test tokens
    regex.lastIndex = offset;
    const match = regex.exec(dataset);
    if (!match) {
      if (dataset.length-offset == 0) return;
      throw new Error(`invalid syntax at line ${this.line} col ${this.col}:`);
    }
    const [ text, ...groups ] = match;
    // Determine the rule
    let rule, i;
    for (i = 0; i < rules.length; i++) {
      if (groups[i]) rule = rules[i];
    }
    if (!rule) throw new Error('could not find token for match');
    // Assemble node
    const token: Token = {
      type: rule.id,
      value: rule.value ? rule.value(text) : text,
      text: text,
      offset: offset,
      line: line,
      col: col
    };
    // Keep track of position
    this.offset += text.length;
    this.col += text.length;
    if (rule.lineBreaks) {
      // Check if contains lineBreaks and count
      const a: number[] = [];
      let i=-1;
      while((i=text.indexOf('\n',i+1)) >= 0) a.push(i);
      if (a.length != 0) {
        this.line += a.length;
        this.col = text.length - <number>a.pop();
      }
    }
    return token;
  }
  save() {
    return {
      offset: this.offset,
      line: this.line,
      col: this.col
    };
  }
  reset(chunk: string, info?: { offset: number, line: number, col: number }) {
    this.dataset = chunk;
    this.offset = info ? info.offset : 0;
    this.line = info ? info.line : 1;
    this.col = info ? info.col : 1;
  }
  formatError(token: Token) {
    if (token == null) return 'parseError: at end of file';
    return `parseError: at line ${token.line} col ${token.col}:`;
  }
  has(name: string) {
    return this.rules.some(({ id }) => id == name);
  }
}
export default () => new Lexer(Tokens);