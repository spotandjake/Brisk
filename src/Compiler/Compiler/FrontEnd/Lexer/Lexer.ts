import rules from './Tokens';
import { BriskSyntaxError } from '../../../Errors/Compiler';
// Import Types
import { Token, Lexeme } from '../../Types';
class Lexer {
  public file: string;
  private regex: RegExp;
  private rules: Lexeme[] = rules;
  private dataset = '';
  private offset = 0;
  private line = 1;
  private col = 1;
  constructor(file: string) {
    // store stuff
    this.regex = new RegExp(`(?:${
      rules.map(({ match }) => {
        if (match.ignoreCase || match.global || match.sticky || match.multiline)
          BriskSyntaxError('RegExp /i/g/y/m flag used');
        return `((?:${match.source}))`;
      }).join('|')})`, 'my');
    this.file = file;
  }
  next(): (Token | undefined) {
    const { regex, rules, dataset, offset, line, col, file } = this;
    // Test tokens
    regex.lastIndex = offset;
    const match = regex.exec(dataset);
    if (!match) {
      if (dataset.length-offset == 0) return;
      BriskSyntaxError(`invalid syntax at line ${line} col ${col}:`);
      process.exit(1); // Tricks Lsp
    }
    const [ text, ...groups ] = match;
    // Determine the rule
    const rule = rules.find((r, i) => groups[i]);
    if (!rule) {
      BriskSyntaxError('could not find token for match');
      process.exit(1); // Tricks Lsp
    }
    // Keep track of position
    this.offset += text.length;
    this.col += text.length;
    if (rule.lineBreaks) {
      // Check if contains lineBreaks and count
      const a: number[] = [];
      let i=-1;
      while((i=text.indexOf('\n',i+1)) >= 0) a.push(i);
      if (a.length != 0) {
        this.line += a.length;
        this.col = text.length - <number>a.pop();
      }
    }
    return {
      type: rule.id,
      value: rule.value ? rule.value(text) : text,
      text: text,
      offset: offset,
      line: line,
      col: col,
      file: file
    };
  }
  save() {
    return {
      offset: this.offset,
      line: this.line,
      col: this.col,
      file: this.file
    };
  }
  reset(chunk: string, info?: { offset: number, line: number, col: number, file: string }) {
    this.dataset = chunk;
    this.offset = info?.offset || 0;
    this.line = info?.line || 1;
    this.col = info?.col || 1;
    this.file = info?.file || this.file;
  }
  formatError(token: Token) {
    return `parseError: ${token == null ? 'at end of file' : `at line ${token.line} col ${token.col}:`}`;
  }
  has(name: string) {
    return this.rules.some(({ id }) => id == name);
  }
}
export default Lexer;