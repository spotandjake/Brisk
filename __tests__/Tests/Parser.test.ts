// Test Utils
import { test, expect } from '@jest/globals';
import { ILexingResult } from 'chevrotain';
import fs from 'fs';
// Test Components
import { Tokens } from '../../src/Compiler/Lexer/Tokens';
import parse from '../../src/Compiler/Parser/index';
// Import Data: Relative to dist
// Lexer Tests
test('Parser: Main Pass', () => {
  const data = JSON.parse(fs.readFileSync('./__tests__/Data/Parser/Parser_Pass.json', 'utf8'));
  // console.log(Tokens);
  const Parser_Main: ILexingResult = {
    ...data,
    tokens: data.tokens.map((tkn: any) => {
      return {
        ...tkn,
        tokenType: Tokens.find((token) => token.name === tkn.tokenType),
      };
    })
  };
  expect(parse(Parser_Main, 'file')).toMatchSnapshot();
});
test('Parser: Literals Pass', () => {
  const data = JSON.parse(fs.readFileSync('./__tests__/Data/Parser/Literals_Pass.json', 'utf8'));
  // console.log(Tokens);
  const Parser_Main: ILexingResult = {
    ...data,
    tokens: data.tokens.map((tkn: any) => {
      return {
        ...tkn,
        tokenType: Tokens.find((token) => token.name === tkn.tokenType),
      };
    })
  };
  expect(parse(Parser_Main, 'file')).toMatchSnapshot();
});