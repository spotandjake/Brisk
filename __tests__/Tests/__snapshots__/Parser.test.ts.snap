// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser: Main 1`] = `
Object {
  "body": Array [
    Object {
      "category": 1,
      "nodeType": 4,
      "position": Object {
        "col": 1,
        "file": "file",
        "line": 2,
        "offset": 19,
      },
      "source": Object {
        "category": 3,
        "nodeType": 15,
        "position": Object {
          "col": 15,
          "file": "file",
          "line": 2,
          "offset": 33,
        },
        "value": "'test'",
      },
      "variable": Object {
        "category": 5,
        "name": "a",
        "nodeType": 21,
        "position": Object {
          "col": 8,
          "file": "file",
          "line": 2,
          "offset": 26,
        },
      },
    },
    Object {
      "category": 1,
      "nodeType": 5,
      "position": Object {
        "col": 1,
        "file": "file",
        "line": 3,
        "offset": 42,
      },
      "source": Object {
        "category": 3,
        "nodeType": 15,
        "position": Object {
          "col": 42,
          "file": "file",
          "line": 3,
          "offset": 83,
        },
        "value": "'print'",
      },
      "typeSignature": Object {
        "category": 4,
        "name": "Function",
        "nodeType": 19,
        "params": Array [
          Object {
            "category": 4,
            "name": "String",
            "nodeType": 20,
            "position": Object {
              "col": 21,
              "file": "file",
              "line": 3,
              "offset": 62,
            },
          },
        ],
        "position": Object {
          "col": 20,
          "file": "file",
          "line": 3,
          "offset": 61,
        },
        "returnType": Object {
          "category": 4,
          "name": "Void",
          "nodeType": 20,
          "position": Object {
            "col": 32,
            "file": "file",
            "line": 3,
            "offset": 73,
          },
        },
      },
      "variable": Object {
        "category": 5,
        "name": "print",
        "nodeType": 21,
        "position": Object {
          "col": 13,
          "file": "file",
          "line": 3,
          "offset": 54,
        },
      },
    },
    Object {
      "category": 1,
      "nodeType": 5,
      "position": Object {
        "col": 1,
        "file": "file",
        "line": 4,
        "offset": 93,
      },
      "source": Object {
        "category": 3,
        "nodeType": 15,
        "position": Object {
          "col": 32,
          "file": "file",
          "line": 4,
          "offset": 124,
        },
        "value": "'print'",
      },
      "typeSignature": Object {
        "category": 4,
        "name": "Function",
        "nodeType": 19,
        "params": Array [],
        "position": Object {
          "col": 16,
          "file": "file",
          "line": 4,
          "offset": 108,
        },
        "returnType": Object {
          "category": 4,
          "name": "Void",
          "nodeType": 20,
          "position": Object {
            "col": 22,
            "file": "file",
            "line": 4,
            "offset": 114,
          },
        },
      },
      "variable": Object {
        "category": 5,
        "name": "f",
        "nodeType": 21,
        "position": Object {
          "col": 13,
          "file": "file",
          "line": 4,
          "offset": 105,
        },
      },
    },
    Object {
      "category": 1,
      "nodeType": 5,
      "position": Object {
        "col": 1,
        "file": "file",
        "line": 5,
        "offset": 134,
      },
      "source": Object {
        "category": 3,
        "nodeType": 15,
        "position": Object {
          "col": 25,
          "file": "file",
          "line": 5,
          "offset": 158,
        },
        "value": "'test'",
      },
      "typeSignature": Object {
        "category": 4,
        "name": "i32",
        "nodeType": 20,
        "position": Object {
          "col": 16,
          "file": "file",
          "line": 5,
          "offset": 149,
        },
      },
      "variable": Object {
        "category": 5,
        "name": "b",
        "nodeType": 21,
        "position": Object {
          "col": 13,
          "file": "file",
          "line": 5,
          "offset": 146,
        },
      },
    },
    Object {
      "alternative": Object {
        "alternative": Object {
          "body": Array [
            Object {
              "args": Array [
                Object {
                  "category": 3,
                  "nodeType": 15,
                  "position": Object {
                    "col": 9,
                    "file": "file",
                    "line": 14,
                    "offset": 354,
                  },
                  "value": "'Else Test'",
                },
              ],
              "category": 2,
              "name": Object {
                "category": 5,
                "name": "print",
                "nodeType": 22,
                "position": Object {
                  "col": 3,
                  "file": "file",
                  "line": 14,
                  "offset": 348,
                },
              },
              "nodeType": 13,
              "position": Object {
                "col": 3,
                "file": "file",
                "line": 14,
                "offset": 348,
              },
            },
          ],
          "category": 1,
          "nodeType": 3,
          "position": Object {
            "col": 8,
            "file": "file",
            "line": 13,
            "offset": 343,
          },
        },
        "body": Object {
          "body": Array [
            Object {
              "args": Array [
                Object {
                  "category": 3,
                  "nodeType": 15,
                  "position": Object {
                    "col": 9,
                    "file": "file",
                    "line": 12,
                    "offset": 318,
                  },
                  "value": "'Else If Test'",
                },
              ],
              "category": 2,
              "name": Object {
                "category": 5,
                "name": "print",
                "nodeType": 22,
                "position": Object {
                  "col": 3,
                  "file": "file",
                  "line": 12,
                  "offset": 312,
                },
              },
              "nodeType": 13,
              "position": Object {
                "col": 3,
                "file": "file",
                "line": 12,
                "offset": 312,
              },
            },
          ],
          "category": 1,
          "nodeType": 3,
          "position": Object {
            "col": 20,
            "file": "file",
            "line": 11,
            "offset": 307,
          },
        },
        "category": 1,
        "condition": Object {
          "category": 3,
          "nodeType": 15,
          "position": Object {
            "col": 12,
            "file": "file",
            "line": 11,
            "offset": 299,
          },
          "value": "'test'",
        },
        "nodeType": 1,
        "position": Object {
          "col": 8,
          "file": "file",
          "line": 11,
          "offset": 295,
        },
      },
      "body": Object {
        "body": Array [
          Object {
            "args": Array [
              Object {
                "category": 3,
                "nodeType": 15,
                "position": Object {
                  "col": 9,
                  "file": "file",
                  "line": 10,
                  "offset": 275,
                },
                "value": "'If Test'",
              },
            ],
            "category": 2,
            "name": Object {
              "category": 5,
              "name": "print",
              "nodeType": 22,
              "position": Object {
                "col": 3,
                "file": "file",
                "line": 10,
                "offset": 269,
              },
            },
            "nodeType": 13,
            "position": Object {
              "col": 3,
              "file": "file",
              "line": 10,
              "offset": 269,
            },
          },
        ],
        "category": 1,
        "nodeType": 3,
        "position": Object {
          "col": 13,
          "file": "file",
          "line": 9,
          "offset": 264,
        },
      },
      "category": 1,
      "condition": Object {
        "category": 2,
        "lhs": Object {
          "category": 3,
          "nodeType": 16,
          "position": Object {
            "col": 5,
            "file": "file",
            "line": 9,
            "offset": 256,
          },
          "value": "1",
        },
        "nodeType": 9,
        "operator": 0,
        "position": Object {
          "col": 5,
          "file": "file",
          "line": 9,
          "offset": 256,
        },
        "rhs": Object {
          "category": 3,
          "nodeType": 16,
          "position": Object {
            "col": 10,
            "file": "file",
            "line": 9,
            "offset": 261,
          },
          "value": "1",
        },
      },
      "nodeType": 1,
      "position": Object {
        "col": 1,
        "file": "file",
        "line": 9,
        "offset": 252,
      },
    },
    Object {
      "alternative": undefined,
      "body": Object {
        "args": Array [
          Object {
            "category": 3,
            "nodeType": 15,
            "position": Object {
              "col": 19,
              "file": "file",
              "line": 16,
              "offset": 391,
            },
            "value": "'test'",
          },
        ],
        "category": 2,
        "name": Object {
          "category": 5,
          "name": "print",
          "nodeType": 22,
          "position": Object {
            "col": 13,
            "file": "file",
            "line": 16,
            "offset": 385,
          },
        },
        "nodeType": 13,
        "position": Object {
          "col": 13,
          "file": "file",
          "line": 16,
          "offset": 385,
        },
      },
      "category": 1,
      "condition": Object {
        "category": 2,
        "lhs": Object {
          "category": 3,
          "nodeType": 16,
          "position": Object {
            "col": 5,
            "file": "file",
            "line": 16,
            "offset": 377,
          },
          "value": "1",
        },
        "nodeType": 9,
        "operator": 0,
        "position": Object {
          "col": 5,
          "file": "file",
          "line": 16,
          "offset": 377,
        },
        "rhs": Object {
          "category": 3,
          "nodeType": 16,
          "position": Object {
            "col": 10,
            "file": "file",
            "line": 16,
            "offset": 382,
          },
          "value": "1",
        },
      },
      "nodeType": 1,
      "position": Object {
        "col": 1,
        "file": "file",
        "line": 16,
        "offset": 373,
      },
    },
  ],
  "category": 0,
  "nodeType": 0,
  "position": Object {
    "col": 0,
    "file": "file",
    "line": 0,
    "offset": 0,
  },
}
`;
